torchsparse/backend/convolution/convolution_cuda.cu

```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
+                     neighbor_offset.data_ptr<int>() + kernel_volume));
+    max_kmap_size = std::max(max_kmap_size, 1);
+    // (N, c) X (c, o) = (N, o)
+    // torch::mm_out(out_feat, in_feat, kernel[mid_kernel]);
+  } else {
+    max_kmap_size =
+        *std::max_element(neighbor_offset.data_ptr<int>(),
+                          neighbor_offset.data_ptr<int>() + kernel_volume);
+  }
+
+  std::vector<std::vector<int>> groups;
+  std::vector<int> mm_ops;
+  std::vector<int> group_sizes;
+  at::Tensor pads, cum_pads;
+  int buffer_size;
+  // step0.1: get the groups
+  group_strategy_generation(kernel_volume, epsilon, mm_thresh, conv_mode,
+                            neighbor_offset, precompute_mid, groups, mm_ops,
+                            group_sizes, cum_buffer_sizes, buffer_size);
+  at::Tensor cum_buffer_sizes_gpu =
+      cum_buffer_sizes.to(neighbor_offset_cum_gpu.device());
+
+  // symmetric_mode &= precompute_mid;
+  at::Tensor in_buffer, out_buffer;
+  if (!in_feat.requires_grad()) {
+    if (is_half) {
+      in_buffer = torch::from_blob(global_buffer.data_ptr<at::Half>(),
+                                   {buffer_size, n_in_channels}, options);
+      out_buffer = torch::from_blob(
+          global_buffer.data_ptr<at::Half>() + buffer_size * n_in_channels,
+          {buffer_size, n_out_channels}, options);
+    } else {
+      in_buffer = torch::from_blob(global_buffer.data_ptr<float>(),
+                                   {buffer_size, n_in_channels}, options);
+      out_buffer = torch::from_blob(
+          global_buffer.data_ptr<float>() + buffer_size * n_in_channels,
+          {buffer_size, n_out_channels}, options);
+    }
+  } else {
+    in_buffer = torch::zeros({buffer_size, n_in_channels}, options);
+    out_buffer = torch::zeros({buffer_size, n_out_channels}, options);
+  }
+
+  // all gather
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      in_feat.type(), "convolution_forward_cuda", ([&] {
+        gather_all_kernel_pad_sep_with_mask<scalar_t>
+            <<<ceil((double)(n_in_feats * n_in_channels) /
+                    (256 << (sizeof(scalar_t) == 2) + 2)),
+               256>>>(n_in_feats, n_in_channels, kernel_volume,
+                      in_feat.data_ptr<scalar_t>(),
+                      in_buffer.data_ptr<scalar_t>(),
+                      neighbor_map.data_ptr<int>(),
+                      neighbor_offset_gpu.data_ptr<int>(),
+                      neighbor_offset_cum_gpu.data_ptr<int>(),
+                      cum_buffer_sizes_gpu.data_ptr<int>(),
+                      input_mask.data_ptr<int>(), output_mask.data_ptr<int>(),
+                      transpose, precompute_mid);
+      }));
+
+  at::Tensor in_buffer_activated, out_buffer_activated, kernel_buffer;
+  int buffer_st;
+  int cur_buffer_size;
+  // for each group
+  // mm_ops = 0, sep; mm_ops = 1, BMM
+  int kernel_cnt = 0;
+  for (int i = 0; i < groups.size(); i++) {
+    // case 0 and 1: no padding
```

Comment:
Let's remove this comment, it is related to our old implementation and is confusing for the current one (`mm_ops=1` usually means that there are paddings).